internal/zinc-apiinfo/src/main/scala/sbt/internal/inc/ClassToAPI.scala

```diff
@@ -90,33 +90,29 @@ object ClassToAPI {
    *
    * In order to fully fix this issue, the Scala compiler should do the desired behaviour
    * in 2.13.x. 2.12.x is already lost since this fix is binary incompatible. The appropriate
-   * JIRA ticket is here: https://github.com/scala/bug/issues/2034.
+   * issue is here: https://github.com/scala/bug/issues/2034.
    *
-   * @return The canonical name if not null, the full name otherwise.
+   * @return The canonical name if not null, the blank string otherwise.
    */
-  def handleMalformedNameOf(c: Class[_], isRecursive: Boolean = false): String = {
+  def classCanonicalName(c: Class[_]): String = {
     if (c == null) "" // Return nothing if it hits the top-level class
     else {
       val className = c.getName
-      // Adds a `dollar` if it's a recursive call, else nothing
-      val atEnd: String = if (isRecursive) "$" else ""
       try {
         val canonicalName = c.getCanonicalName
         if (canonicalName == null) className
-        else canonicalName + atEnd
+        else canonicalName
       } catch {
         case malformedError: java.lang.InternalError
             if malformedError.getMessage.contains("Malformed class name") =>
           val enclosingClass = c.getEnclosingClass
           val enclosingName = enclosingClass.getName
           val restOfName = c.getName.stripPrefix(enclosingName)
-          handleMalformedNameOf(enclosingClass, true) + restOfName + atEnd
+          classCanonicalName(enclosingClass).stripSuffix("$") + "." + restOfName
```

Comment:
There's a strong difference between my previous PR and yours. Mine was not pretending that the borked binary names had `$$`, it was instead using the original `$` separation to have a consistent way to represent these forked names.
  
The patch that you propose does pretend that the borked binary names are valid. There's nothing wrong with this a priori, but needs a deeper study.
  
For example, what happens if there indeed exists a valid binary name `A$$B$` (it could come from a future version of Scala or another JVM-based programming language)? IT means we'll have a duplicated representation for the borked and the non-borked one.